Description: Fix C++23 incomplete type error in inspector protocol templates
 C++23 with libc++ requires complete type definition when std::unique_ptr
 destructor is instantiated. The inspector protocol generator templates
 generate inline destructors for protocol types, but these types may have
 std::unique_ptr members pointing to forward-declared types.
 .
 Fix by moving destructor definitions from header (.h) to implementation
 (.cpp) templates. The .cpp file includes the header which provides
 complete type definitions for all types in the namespace.
 .
 This fixes errors like:
   error: invalid application of 'sizeof' to an incomplete type
   'v8_inspector::protocol::Runtime::ObjectPreview'
Author: Juan Manuel Mendez Rey <vejeta@gmail.com>
Bug-Debian: https://bugs.debian.org/915400
Forwarded: not-needed
Last-Update: 2025-11-26

diff --git a/third_party/inspector_protocol/templates/TypeBuilder_cpp.template b/third_party/inspector_protocol/templates/TypeBuilder_cpp.template
index a38182c..c637eeb 100644
--- a/third_party/inspector_protocol/templates/TypeBuilder_cpp.template
+++ b/third_party/inspector_protocol/templates/TypeBuilder_cpp.template
@@ -55,6 +55,9 @@ const char* {{type.id}}::{{property.name | to_title_case}}Enum::{{literal | dash
       {% endif %}
     {% endfor %}
     {% if not (type.type == "object") or not ("properties" in type) %}{% continue %}{% endif %}
+// C++23 requires complete types for std::unique_ptr destructor
+{{type.id}}::~{{type.id}}() = default;
+
 CRDTP_BEGIN_DESERIALIZER({{type.id}})
       {% for property in type.properties | sort(attribute = 'name', case_sensitive=True) %}
       {% if property.optional %}
diff --git a/third_party/inspector_protocol/templates/TypeBuilder_h.template b/third_party/inspector_protocol/templates/TypeBuilder_h.template
index 7c2a696..ad2e549 100644
--- a/third_party/inspector_protocol/templates/TypeBuilder_h.template
+++ b/third_party/inspector_protocol/templates/TypeBuilder_h.template
@@ -74,7 +74,8 @@ namespace {{param.name | to_title_case}}Enum {
 class {{config.protocol.export_macro}} {{type.id}} : public ::{{config.crdtp.namespace}}::ProtocolObject<{{type.id}}>{% if protocol.is_exported(domain.domain, type.id) %},
     public API::{{type.id}}{% endif %} {
 public:
-    ~{{type.id}}() override { }
+    // C++23 requires destructor definition in .cpp where types are complete
+    ~{{type.id}}() override;
     {% for property in type.properties %}
       {% set property_type = protocol.resolve_type(property) %}
       {% set property_name = property.name | to_title_case %}
